src/server/pfs/server/local_block_api_server.go

```diff
@@ -48,9 +49,25 @@ func (s *localBlockAPIServer) PutBlock(putBlockServer pfsclient.BlockAPI_PutBloc
 	result := &pfsclient.BlockRefs{}
 	defer func(start time.Time) { s.Log(nil, result, retErr, time.Since(start)) }(time.Now())
 	defer drainBlockServer(putBlockServer)
-	reader := bufio.NewReader(protostream.NewStreamingBytesReader(putBlockServer))
+
+	putBlockRequest, err := putBlockServer.Recv()
+	if err != nil {
+		if err != io.EOF {
+			return err
+		}
+		// Allow empty PutBlock requests, in this case we don't create any actual blockRefs
+		return putBlockServer.SendAndClose(result)
+	}
+
+	reader := bufio.NewReader(&putBlockReader{
+		server: putBlockServer,
+		buffer: bytes.NewBuffer(putBlockRequest.Value),
+	})
+
+	decoder := json.NewDecoder(reader)
+
 	for {
-		blockRef, err := s.putOneBlock(reader)
+		blockRef, err := s.putOneBlock(putBlockRequest.Delimiter, reader, decoder)
```

Comment:
Originally that's what I was doing ... but in this case decoder gets destroyed across calls to `putOneBlock` which means that the decoder loses track of how much its read. This results in json content past the first block getting mangled.